presto-hive/src/main/java/com/facebook/presto/hive/util/MergingPageIterator.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.util;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.SortOrder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+import static com.facebook.presto.hive.util.TempFileWriter.appendPositionTo;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class MergingPageIterator
```

Comment:
this problem is already solved in `MergeHashSort`. I wonder if we could extract common code into separate module. There are few merging places already:
* aggregation pages merge (also from files)
* distributed sort (to be landed), https://github.com/prestodb/presto/pull/9854/files#diff-f156fb513b871f7dc6a36b492d686bc8R196
* now hive sorted buckets writer (yet with an another slightly different approach).

Reply:
I saw that, but it seemed rather complex and not available as a library. At some point, we need to rework the `finish()` methods of `ConnectorPageSink` and `UpdatablePageSource` to return a resumable work object (and make implementations work in chunks).

Reply:
> I saw that, but it seemed rather complex and not available as a library.

I didn't move it to a separate module because there was no need at that time. It should be fairy easy to extract.

> At some point, we need to rework the finish() methods of ConnectorPageSink and UpdatablePageSource to return a resumable work object (and make implementations work in chunks).

What's the granularity of `finish()` call for this code. Single merged page is being written or it is whole data written to disk?

`MergeHashSort` is exactly the same problem. Merge rows from multiple files into a single output page.